crates/evm/src/precompile_cache.rs

```diff
@@ -0,0 +1,341 @@
+                            entry.result = Ok(result.clone());
+                            entry.upper_gas_limit = entry.upper_gas_limit.min(gas_limit);
+                        } else {
+                            cache.cache.insert(
+                                key,
+                                CacheEntry {
+                                    result: Ok(result.clone()),
+                                    upper_gas_limit: gas_limit,
+                                    lower_gas_limit: 0,
+                                },
+                            );
+                        }
+                    } else {
+                        // for other errors cache the result
+                        cache.cache.insert(
+                            key,
+                            CacheEntry {
+                                result: Ok(result.clone()),
+                                upper_gas_limit: gas_limit,
+                                lower_gas_limit: 0,
+                            },
+                        );
+                    }
+
+                    #[cfg(feature = "metrics")]
+                    {
+                        self.metrics.precompile_cache_misses.increment(1);
+                        let new_entry_count = cache.cache.entry_count();
+                        if new_entry_count > entry_count_before {
+                            self.metrics
+                                .precompile_cache_size
+                                .increment((new_entry_count - entry_count_before) as f64);
+                        }
+                    }
+                }
+                Err(err) => {
+                    // fatal error
+                    #[cfg(feature = "metrics")]
+                    let entry_count_before = cache.cache.entry_count();
+
+                    cache.cache.insert(
+                        key,
+                        CacheEntry {
+                            result: Err(err.clone()),
+                            upper_gas_limit: gas_limit,
+                            lower_gas_limit: 0,
+                        },
+                    );
+
+                    #[cfg(feature = "metrics")]
+                    {
+                        self.metrics.precompile_cache_misses.increment(1);
+                        let new_entry_count = cache.cache.entry_count();
+                        if new_entry_count > entry_count_before {
+                            self.metrics
+                                .precompile_cache_size
+                                .increment((new_entry_count - entry_count_before) as f64);
+                        }
+                    }
+                }
+                Ok(None) => {
+                    // precompile not found in inner provider
+                }
+            }
+
+            output
+        } else {
+            self.precompile_provider.run(context, address, inputs, is_static, gas_limit)
+        }
+    }
+
+    #[cfg(not(feature = "std"))]
```

Comment:
so this has feature gated run functions?

how will this be enabled in the actual evm based on the cli arg?

Reply:
yep, this is the no-std version of the run method, just no cache and fallback to the inner precompile provider. i added it as a separate fn to prevent adding more feature checks on the main one. so, to your question, on no-std the flag to enable precompile cache has no effect, lmk if we should change that